### PR TITLE
AddAzureBicepPostgres(...).AddDatabase(...) => AddPostgres(...).PublishAsAzurePostgresFlexibleServer(...).AddDatabase(...)

### PR DESCRIPTION
**Pull request by mitchdenny**
- [x] Renames `AzureBicepPostgresResource` to `AzurePostgresResource'
- [x] Removes `AddAzureBicepPostgres(...)` and adds PublishAsAzurePostgresFlexibleServer(...)` and `AsAzurePostgresFlexibleServer(...)`
- [x] Adds support for deferred evaluation of parameters with `WithParameter(...)`.
- [x] Turns `AzurePostgresResource` into a surrogate for `PostgresServerResource`.
- [x] Changes `IResourceBuilder<PostgresDatabaseResource>.AddDatabase(...)` to emit `value.v0` resource.

### CODE REVIEW

File: playground/bicep/BicepSample.AppHost/aspire-manifest.json [hunk 1]
```diff
@@ -99,17 +108,16 @@
...
           "db2"
-        ],
-        "keyVaultName": ""
+        ]
       }
     },
     "db2": {
-      "type": "azure.bicep.v0",
-      "connectionString": "{postgres2.connectionString};Database=db2",
+      "type": "postgres.database.v0",
```
Comment (davidfowl) [inline playground/bicep/BicepSample.AppHost/aspire-manifest.json:120]: We want the connection string expressed in the manifest

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs [hunk 2]
```diff
@@ -30,6 +31,11 @@ public sealed class ManifestPublishingContext(DistributedApplicationExecutionCon
     /// </summary>
     public Utf8JsonWriter Writer { get; } = writer;
 
+    /// <summary>
+    /// The distributed application model.
+    /// </summary>
+    public DistributedApplicationModel AppModel { get; } = appModel;
```
Comment (davidfowl) [inline src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:37]: I think we should remove this.

File: src/Aspire.Hosting.Azure/Postgres/AzurePostgresResource.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    public override string Name => innerResource.Name;
+
+    /// <inheritdoc />
+    public override ResourceMetadataCollection Annotations => innerResource.Annotations;
+
+    /// <inheritdoc />
+    public override void WriteToManifest(ManifestPublishingContext context)
+    {
+        var databaseList = context.AppModel.Resources.OfType<PostgresDatabaseResource>().Where(db => db.Parent == innerResource).Select(db => db.Name).ToList();
+        Parameters.Add("databases", databaseList);
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Postgres/AzurePostgresResource.cs:48]: Why is this here?

File: src/Aspire.Hosting.Azure/Extensions/AzurePostgresExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        var resource = builder.Resource;
+        return builder.WithManifestPublishingCallback(resource.WriteToManifest)
+                      .WithParameter("serverName", resource.CreateBicepResourceName())
+                      .WithParameter(AzureBicepResource.KnownParameters.KeyVaultName)
+                      .WithParameter("databases", () =>
+                      {
+                          // Evaluation of databases to bind to parameters is deferred until
+                          // the parameter is requested.
+                          return builder.ApplicationBuilder.Resources.OfType<PostgresDatabaseResource>().Where(db => db.Parent == builder.Resource.InnerResource).Select(db => db.Name);
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Extensions/AzurePostgresExtensions.cs:106]: resource.Databases.Select(d => d.Name);

File: src/Aspire.Hosting.Azure/Postgres/AzurePostgresResource.cs (hunk 3 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.Azure/Postgres/AzurePostgresResource.cs:48]: Going away in the next push

File: src/Aspire.Hosting.Azure/Extensions/AzurePostgresExtensions.cs (hunk 4 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.Azure/Extensions/AzurePostgresExtensions.cs:106]: Much simpler now.

File: playground/bicep/BicepSample.AppHost/aspire-manifest.json (hunk 1 above)
Comment (mitchdenny) [reply playground/bicep/BicepSample.AppHost/aspire-manifest.json:120]: I haven't figured out how to do this yet without replacing the manifest publisher on the child resource. The intuitive thing to do is that in the manifest publishing callback for `postgres.database.v0` we just call `GetConnectionString()` ... we redirect to `AzurePostgresResource.GetConnectionString()` but that `GetConnectionString()` call doesn't know we are in manifest publishing mode and tries to return the actual connection string, not the manifest placeholder value.

Comment (mitchdenny) [reply playground/bicep/BicepSample.AppHost/aspire-manifest.json]: This is now `value.v0`.

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Hosting.Azure/Postgres/AzurePostgresResource.cs [hunk 5]
```diff
@@ -44,9 +44,6 @@ public class AzurePostgresResource(PostgresServerResource innerResource) :
     /// <inheritdoc />
     public override void WriteToManifest(ManifestPublishingContext context)
     {
-        var databaseList = context.AppModel.Resources.OfType<PostgresDatabaseResource>().Where(db => db.Parent == innerResource).Select(db => db.Name).ToList();
-        Parameters.Add("databases", databaseList);
-
         base.WriteToManifest(context);
     }
```
[the remark this answers is not in the document]
Comment (mitchdenny) [reply src/Aspire.Hosting.Azure/Postgres/AzurePostgresResource.cs]: Strange you must have been looking at an old version of the code because I removed that last night and pushed it.

[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Postgres/AzurePostgresResource.cs]: Yep I think so too.
